### PR TITLE
Add systemd service to bitcoind in debian package

### PR DESCRIPTION
**Pull request by ghost**
On suggestion from @TheBlueMatt I have updated `contrib/debian` files to include a systemd service in the `bitcoind` build. Tested and working on Ubuntu 16.04 and 17.10.

This fixes Issue #12758

### CODE REVIEW

**Review by randolf — APPROVED** [review]

**Comment by randolf** [comment]
@ctp-tsteenholdt Please remember to squash/rebase.

**Comment by ctp-tsteenholdt** [comment]
@randolf That makes total sense. My apologies - looking into it right now...

**Comment by laanwj** [comment]
We already have a `bitcoind.service` in `contrib/init/` - what is the difference with this one, can we avoid duplicating it?

**Comment by ctp-tsteenholdt** [comment]
@randolf The pull request should look better now - please let me know if I'm missing something.

**Comment by ctp-tsteenholdt** [comment]
@laanwj I'm completely with you on that question, but I don't think we can feasibly avoid multiple versions of a file like this. The different versions of the `bitcoind.service` in the `contrib/` tree have subtle differences to make them work best for their particular purpose. The differences reflect distro-specifics such as location of environment files, dependency services etc.

Also, building a package for Ubuntu or Debian requires a full copy of the `contrib/debian/` folder, including the `bitcoind.service` file, to allow the packaging helper `dh` to recognize, treat and install it correctly.

It think distro-specific files are probably better maintained within distro specific folders such as `contrib/debian/` or `contrib/rpm/`, as part of the overall package maintenance for that distro. This avoids possible breakage from a change made for a different distro too.

To underline that the `contrib/init/` version is still very relevant, it will be the goto file for people on platforms without a specific package or people who prefer to build from source. This file can be easily adapted to work well on most systems.

Please let me know what you think.

**Comment by laanwj** [comment]
@ctp-tsteenholdt If it is intentional, then IMO it's not a problem. Arguably the debian-specific file is easier to maintain because at least it aims for a specific environment and can be tested there, not 'all distributions in general', which has made the file under `contrib/init/` a real pain in practice (see e.g. discussion in #12255).

**Comment by ctp-tsteenholdt** [comment]
@laanwj That's exactly right.

**Comment by laanwj** [comment]
Would be good to have an ACK from @TheBlueMatt here.

File: contrib/debian/bitcoind.postrm
Comment (TheBlueMatt) [inline contrib/debian/bitcoind.postrm]: Maybe make it explicit that they should check /var/lib/bitcoin for things like wallets?

File: contrib/debian/bitcoind.service
Comment (TheBlueMatt) [inline contrib/debian/bitcoind.service]: May be nice to make the datadir explicit so that we dont have /var/lib/bitcoin/.bitcoin and instead just put stuff in /var/lib/bitcoin?

File: contrib/debian/rules
Comment (TheBlueMatt) [inline contrib/debian/rules]: Can you pull the backporting of existing debian/ contents out into a separate commit to make it easier to track?

**Review by TheBlueMatt — COMMENTED** [review]
Looks good, though I didnt test any of it.

**Comment by ctp-tsteenholdt** [comment]
Thanks @TheBlueMatt. Each of your comments make perfect sense and I'll have them all addressed.

**Comment by ctp-tsteenholdt** [comment]
I've addressed the suggestions made by @TheBlueMatt, run a test dpkg-buildpackage using the updated `contrib/debian/` tree and tested that the package changes work as expected, with success.

The commits have been updated to reflect the changes.

**Comment by ctp-tsteenholdt** [comment]
Travis indicates an error but it seems unrelated to this PR?

**Comment by laanwj** [comment]
@TheBlueMatt can you take another look?

**Comment by TheBlueMatt** [comment]
Hmm, if you're gonna pull some of the debian changes upstream, can you just pull all of them? Especially getting the changelog and other stuff out of sync seems not ideal.

File: contrib/debian/bitcoind.postinst
Comment (TheBlueMatt) [inline contrib/debian/bitcoind.postinst]: Why not create the directory? If its empty it will be deleted on uninstall, but if ther user ever tries to use the service things will simply fail to start, no?

Comment (ctp-tsteenholdt) [reply contrib/debian/bitcoind.postinst]: The reason for this approach is that, if for whatever reason the user has changed ownership or permissions on the `/var/lib/bitcoin` directory, we don't want to mess with that on operations like reconfigure, update etc.

`adduser` does not let me specify custom homedir permissions, so I'm using `--no-create-home` to work around that. A missing home directory becomes a stable indicator that this is a fresh install, making it safe for me to create the directory and set the restrictive default permissions.

**Comment by ctp-tsteenholdt** [comment]
@TheBlueMatt I'm not sure what you mean by:

_Hmm, if you're gonna pull some of the debian changes upstream, can you just pull all of them? Especially getting the changelog and other stuff out of sync seems not ideal._

Upstream in this case would essentially be your PPA right? Or is there somewhere else?

I can certainly take a look at any other differences between this tree and yours and see if it makes sense to include in this PR or create a separate one for getting those in sync? Is this what you mean?

File: contrib/debian/bitcoind.postinst
Comment (TheBlueMatt) [reply contrib/debian/bitcoind.postinst]: Oh, I misread the script, sorry.

**Comment by TheBlueMatt** [comment]
Yea, just pull the stuff from the debian PPA tar.

**Comment by ctp-tsteenholdt** [comment]
@TheBlueMatt so I took all the changes from your upstream PPA debian tar and committed them as a single commit. After that, I committed the Systemd changes in a separate commit.

As a precaution I've run a local build and made sure that things still work as expected.

File: contrib/debian/bitcoin-qt.desktop
Comment (fanquake) [inline contrib/debian/bitcoin-qt.desktop]: Note that the additional categories were just added in #12854

Comment (ctp-tsteenholdt) [reply contrib/debian/bitcoin-qt.desktop]: @fanquake thanks. I'll get that sorted out here...

**Comment by ctp-tsteenholdt** [comment]
So I re-resolved the conflict with #12854. This time without nuking the accepted changes of that PR. Sorry about that. Thanks @fanquake for the heads up.

**Comment by laanwj** [comment]
@TheBlueMatt is this now ready for merge?

**Comment by ctp-tsteenholdt** [comment]
Just tested on Ubuntu 18.04, with no issues pertaining to this PR. There is an unrelated issue with the `libdb++` version that needs to be sorted out, so I built with `--with-incompatible-bdb` to test things out.

File: contrib/debian/rules
Comment (TheBlueMatt) [inline contrib/debian/rules]: Why? I dont think anyone would be surprised by their bitcoind going down prior to upgrade instead of only being restarted, and this could break eg a lxc VM as you'll get a text file busy error.

**Review by TheBlueMatt — COMMENTED** [review]
Looks good modulo the one restart question I had.

File: contrib/debian/rules
Comment (ctp-tsteenholdt) [reply contrib/debian/rules]: @TheBlueMatt, good point.

`--restart-after-upgrade` is default in `dh_systemd_start` from compat level 10, so the reason is perhaps best described as emulating modern defaults. I agree that for bitcoind, there should not be any real issues with using the defaults.

I don't quite follow the problems you mention though. As far as I can tell, an upgrade should work perfectly, even when leaving the daemon running during the upgrade. Nothing runtime related is touched during the package upgrade, only the new `bitcoind` and `bitcoin-cli` binaries are put into place, making them ready for the late restart. Unless I'm missing something, this should be no different when running in a container?

If you feel there's a risk here, I'll happily remove this!

Comment (TheBlueMatt) [reply contrib/debian/rules]: Eh, if its default whatever, but linux generally refuses to replace binaries on disk if they are running inside of an LXC/docker container (though will happily do so during normal operation).

**Comment by TheBlueMatt** [comment]
utACK 2a87b1b07c5c4f8b9b34747c5f254c2ae1e824bf with or without restart-after-upgrade.

**Comment by luke-jr** [comment]
How will this affect systems that don't have/use systemd?

**Comment by ctp-tsteenholdt** [comment]
@luke-jr this affects only debian-based systems (including Ubuntu) with systemd.

**Comment by ctp-tsteenholdt** [comment]
Older debian systems that are missing the dh-systemd package, will fail to build the packages, yes. I believe the (supported) Debian/Ubuntu versions that are affected by this, are Debian 7 (LTS version is EOL at the end of May 2018) and Ubuntu 14.04 LTS (EOL in April 2019).

I'm running a test on Ubuntu 14.04 right now, to see what kind of issues we're running into on that platform.

**Comment by ctp-tsteenholdt** [comment]
As it turns out, Ubuntu 14.04 LTS actually does have a `dh-systemd` package and apart from the `libdb4.8++-dev` build dependency, that was also causing problems on Ubuntu 18.04, the build is working just fine.

Since the distro does not actually use systemd, the included `bitcoind.service` simply does nothing.

**Comment by luke-jr** [comment]
I'm less concerned with merely old distros, as much as ones that have simply rejected the mandatory-systemd nonsense, such as Devuan.

**Comment by ctp-tsteenholdt** [comment]
@luke-jr to build the package, you'd need to have `dh-systemd` installed, but you do *not* need `systemd` to install or use it. If you don't have `systemd` on the system, you simply can't use the systemd service configuration that's included here.

**Comment by ctp-tsteenholdt** [comment]
Did that last comment answer your questions?

**Comment by laanwj** [comment]
I'm going to merge this as @TheBlueMatt utACKed this - he's the ppa maintainer. 
I don't think other distributions are relevant here as this concerns explicitly the `debian` directory.
Other improvements can be made later.

**Comment by TheBlueMatt** [comment]
Err, no, Luke's point is good and would break several PPA-supported releases.

**Comment by TheBlueMatt** [comment]
Oh, hmm, I appears to be wrong, it looks like the dh-systemd package *is* on trusty, I suppose its fine, then.

**Comment by laanwj** [comment]
Had to revert this, please file again in 2019 (or whenever relevant).

**Comment by ctp-tsteenholdt** [comment]
I believe dh-systemd is indeed available for all relevant distributions - It certainly is for Devuan, Ubuntu Trusty (and later) and Debian 7 (and later).

**Comment by laanwj** [comment]
We're switching to a different strategy, distribution packaging files will be under this repository from now on:
https://github.com/bitcoin-core/packaging

**Comment by ctp-tsteenholdt** [comment]
Okay... Does this mean I should rework the changes to go in the new path then?

**Comment by MarcoFalke** [comment]
Indeed, just cherry-pick the commits and submit them to the other repo.

**Pull request merged**
